scrapy/utils/misc.py

```diff
@@ -162,3 +164,38 @@ def set_environ(**kwargs):
                 del os.environ[k]
             else:
                 os.environ[k] = v
+
+
+if not six.PY2:
+    from textwrap import dedent
+
+    def is_generator_with_return_value(callable):
+        """
+        Returns True if a callable is a generator function which includes a
+        'return' statement with a value different than None, False otherwise
+        """
+
+        def returns_none(return_node):
+            value = return_node.value
+            return value is None or isinstance(value, ast.NameConstant) and value.value is None
+
+        if inspect.isgeneratorfunction(callable):
+            tree = ast.parse(dedent(inspect.getsource(callable)))
+            for node in ast.walk(tree):
+                if isinstance(node, ast.Return) and not returns_none(node):
+                    return True
+        return False
+
+    def warn_on_generator_with_return_value(spider, callable, logger):
+        """
+        Logs a warning if a callable is a generator function and includes
+        a 'return' statement with a value different than None
+        """
+        if is_generator_with_return_value(callable):
+            logger.warning(
```

Comment:
What do you think about using warnings.warn instead? Otherwise this message will be logged at each request. By using warnings you'll also avoid passing logger argument.

Reply:
Good point. I was passing the logger so the warning would come from `scrapy.core.scraper`, but I guess that's really not relevant. About the multiple warnings, I hadn't realized about that, thanks. Will `warning.warn` be issued only once?

Reply:
yes, warning.warn is only issued once

Reply:
Updated, thanks for the suggestion. Seems like the default class is [`UserWarning`](https://docs.python.org/3/library/exceptions.html#UserWarning), which seems reasonable enough to me according to its description "Base class for warnings generated by user code."